### ISSUE TITLE
Integer overflow in calloc

### ISSUE THREAD

**Issue by ch3root** [TrustInSoft/tis-interpreter#48]
Source code:
```c
#include <stdlib.h>

int main()
{
  char *p = calloc(-2, -3);
  p[5];
  p[6];
}
```

tis-interpreter (3615ce825dfb8a5e66ba60c0e62e9c2131f1cd2d) output:
```
[value] Analyzing a complete application starting at main
[value] Computing initial state
[value] Initial state computed
calloc-overflow.c:7:[kernel] warning: out of bounds read. assert \valid_read(p+6);
                  stack: main
[value] Stopping at nth alarm
[value] user error: Degeneration occurred:
                    results are not correct for lines of code that can be reached from the degeneration point.
```

We see that `p[5]` is fine while `p[6]` is not. Looks like the usual integer overflow in calloc.

Links:
http://cert.uni-stuttgart.de/ticker/advisories/calloc.html
http://kqueue.org/blog/2012/03/05/memory-allocator-security-revisited/

**Comment by pascal-cuoq** [comment]
I have pushed a temporary fix, where the overflow of the product of `calloc`'s arguments (computed as `size_t`) is detected as an error.

Not sure yet that this will be the definite fix. On the one hand, there might exist a program that relies on `calloc(-2, -3)` returning a null pointer for its correct execution. On the other hand, the interpreter only analyzes what happens for successful allocations(*), and there is no successful allocation in this case.

(*) until I resolve https://github.com/TrustInSoft/tis-interpreter/issues/1

**Comment by ch3root** [comment]
If a program calls such `calloc` I think it expects it either to succeed or to fail. The case of overflowed computation inside `calloc` (valid pointer returned but the amount of memory allocated is too small) is something else. In production systems it will lead to vulns. In tis-interpreter it will lead to false positives.

If you think that it's more or less safe to call such `calloc`s with real-world libcs (i.e. that they will return 0) than you can return 0 too. Otherwise it's better to report such `calloc`s.

Another possibility is to report all attempts to allocate more than SIZE_MAX/2 (aka PTRDIFF_MAX) due to the lack of support for such objects in current compilers -- https://gcc.gnu.org/bugzilla/show_bug.cgi?id=67999 . valgrind reports such cases as "Argument 'size' of function malloc has a fishy (possibly negative) value" (maybe this is not related to the bug in compilers, I don't know).
An example of a programmer assuming that malloc will not allocate more than PTRDIFF_MAX -- https://lists.fedorahosted.org/archives/list/elfutils-devel@lists.fedorahosted.org/message/DPKBCHMOWXCT6QBNTD3ERNAC4633OLZ7/ .
OTOH in practice it's only relevant to 32-bit architectures.

**Comment by pascal-cuoq** [comment]
> An example of a programmer assuming that malloc will not allocate more than PTRDIFF_MAX

Hm. There are also plenty of programs that assume that `malloc(0xa0000000)` will succeed in order to handle the quantity of data they need to on 32-bit computers, and that do not invoke UB when the allocation succeeds. It's a shame to exclude some of the behaviors that these programs have in practice with system-provided `malloc`.

(Note that tis-interpreter should warn on subtraction of pointers to the same block that result in a value that overflows `ptrdiff_t`.)

OTOH this question is a bit theoretical because a program that allocates and touches 0xa0000000 bytes of memory is likely to be too slow to execute in tis-interpreter.

**Comment by ch3root** [comment]
Yeah, the situation is complex. The need to handle >=2GB on 32-bit arch is why Florian Weimer doesn't want to "fix" it in glibc.

Issue #9 is exactly from my attempt to see if tis-interpreter warns for overflows in subtraction of pointers. But please note that the referenced gcc bug report (and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=45779) is about miscompilation of valid C code (other than exceeding the (undocumented) object size limit of gcc). The ideal solution would be to fix gcc and clang. Even better would be then define (size_t)(p-q) when p-q overflows ptrdiff_t. But prospects of this are not clear.
